### PR TITLE
Queue, via linked list

### PR DESCRIPTION
**Pull request by EthanJStark**

### CODE REVIEW

File: lib/stack.js [hunk 1]
```diff
@@ -10,4 +10,4 @@ var Stack = function Stack() {
   _classCallCheck(this, Stack);
```
Comment (deonna) [inline lib/stack.js:10]: Consider adding `lib/` to your `.gitignore` to reduce the clutter of autogenerated files in your repo.

File: src/stack.js [hunk 2]
```diff
@@ -1,5 +1,31 @@
 'use strict'
 
 export default class Stack {
```
Comment (deonna) [inline src/stack.js:3]: Initializing `this._topIndex` to `-1` feels unintuitive to me because indices typically start at `0`. 

Instead of starting at `-1` and using a preincrement to increment `this._topIndex` in your `push()` implementation, you can alternatively start at `0` and use a postincrement to increment `this._topIndex`.

This change would simplify your other method implementations.

For example:

```javascript

length() {
  return this._topIndex
}

peek() {
  return this._topIndex ? null : this._storage[ this._topIndex ]
}
```

...etc.

File: src/stack.js [hunk 3]
```diff
@@ -1,5 +1,31 @@
...
-  // your code here
+  constructor(){
+    this._storage = []
```
Comment (deonna) [inline src/stack.js]: Naming: I'd call this something like `elements` instead of `storage` because all variables are essentially storage.

File: src/stack.js [hunk 4]
```diff
@@ -1,5 +1,31 @@
...
+    return this._topIndex === -1 ? null : this._storage[ this._topIndex ]
+  }
+
+  isEmpty() {
+    return this._topIndex === -1
+  }
+
+  pop() {
+    if ( this._topIndex === -1 ) { return null }
+    let result = this._storage[ this._topIndex ]
```
Comment (deonna) [inline src/stack.js]: Use `const` when you're not planning on reinitializing a variable.

File: src/queue.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
+'use strict'
+
+function Node( data ) {
+  this._data = data
+  this._next = null
+}
+
+export default class Queue {
+  constructor() {
+    this._front = null
```
Comment (deonna) [inline src/queue.js:10]: Like the use of `_` to distinguish between instance variables and local variables.

File: src/queue.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  length() {
+    return this._length
+  }
+
+  isEmpty() {
+    return this._length == 0
+  }
+
+  back() {
+    return this._back._data
```
Comment (deonna) [inline src/queue.js]: Error handling: what happens when `this._back` is `null`?

File: src/queue.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  }
+
+  front() {
+    return this._front._data
```
Comment (deonna) [inline src/queue.js]: Error handling: what happens when `this._front` is `null`?

File: src/queue.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+      newNode._next = oldBack
+    }
+    ++this._length
+  }
+
+  dequeue() {
+    let oldFront = this._front
+    --this._length
+    let newFront = this._back
+    for ( let i = 1; i < ( this._length ); ++i) {
```
Comment (deonna) [inline src/queue.js]: Parentheses for `this._length` in a `for` loop are unconventional and a little difficult to parse. Remove for readability.

File: src/queue.js [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+  }
+
+  enqueue( data ) {
+    let newNode = new Node( data )
+    let oldBack = this._back
+    if ( this._length === 0 ) {
+      this._front = newNode
+      this._back = newNode
+    } else {
+      this._back = newNode
```
Comment (deonna) [inline src/queue.js]: This line is in both the `if` and `else` paths in the conditional.

Place it outside the conditional so it runs in both cases and you don't repeat yourself (DRY principle).

File: src/queue.js [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    return this._back._data
+  }
+
+  front() {
+    return this._front._data
+  }
+
+  enqueue( data ) {
+    let newNode = new Node( data )
+    let oldBack = this._back
```
Comment (deonna) [inline src/queue.js]: Declare and initialize `oldBack` in `else` closer to where you're actually using it.

Makes for more readable code when you declare local variables as close as possible to the context in which they're used.

Also, use `const` when you don't plan on reinitializing a variable.

File: spec/queue.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+  context('enqueue()', () => {
+    it('adds an element to the back of the queue', () => {
+      const myQueue = new Queue()
+      myQueue.enqueue( 'foo' )
+      myQueue.enqueue( 'bar' )
+
+      expect( myQueue.back() ).to.eql( 'bar' )
+    })
+    it('increments the length', ()=> {
```
Comment (deonna) [inline spec/queue.js:23]: Add spaces between your tests for readability.

File: spec/queue.js [hunk 12]
```diff
@@ -0,0 +1,60 @@
+import chai, { expect } from 'chai'
+import chaiChange from 'chai-change'
+import Queue from '../src/queue'
+
+chai.use(chaiChange)
+
+describe.only('Queue', () => {
```
Comment (deonna) [inline spec/queue.js]: Add explicit tests for `front()` and `back()` (as well as their expected behavior when the queue is empty.

File: spec/stack.js [hunk 13]
```diff
@@ -12,11 +12,74 @@ describe('Stack', () => {
...
+      const myStack = new Stack()
+      expect( myStack.pop() ).to.be.null
+    })
+    it('if called on a stack with length=1, length becomes 0', () => {
+      const myStack = new Stack()
+      myStack.push('foo')
+      expect(() => myStack.pop())
+        .to.alter(() => myStack.length(), { from: 1, to: 0 })
+    })
+    it('strange case array delete', () => {
```
Comment (deonna) [inline spec/stack.js]: Better description: what's strange about this deletion?

File: spec/stack.js [hunk 14]
```diff
@@ -12,11 +12,74 @@ describe('Stack', () => {
...
+      myStack.pop()
+      expect(myStack.peek()).to.eql('foo')
+    })
+  })
+  context('peek()', () => {
+    it('returns the item at the top of the stack', () => {
+      const myStack = new Stack()
+      myStack.push('foo')
+      myStack.push('bar')
+      expect( myStack.peek() ).to.eql('bar')
```
Comment (deonna) [inline spec/stack.js]: To `equal` - `eql` is for deep equality comparisons (e.g., comparing an array to another array).

Also: `eql` is poorly named by Chai. Use the alternative `to.deep.equal` instead.

**Review by deonna — COMMENTED** [review]
See comments for areas of improvement.

Good job with cleanliness of code and writing good tests.
